R/xgx_stat_ci.R

```diff
@@ -99,87 +99,150 @@
 #' @importFrom binom binom.exact
 #' @importFrom ggplot2 stat_summary
 #' @importFrom ggplot2 aes
-#' @importFrom ggplot2 position_dodge
+#' @importFrom ggplot2 position_dodeg
 #' @export
-xgx_stat_ci <- function(mapping = NULL, data = NULL, conf_level = 0.95,
-                        distribution = "normal",
-                        geom = list("point", "line", "errorbar"),
-                        position = "identity",
-                        fun.args = list(),
-                        na.rm = FALSE,
-                        show.legend = NA,
-                        inherit.aes = TRUE,
-                        ...) {
-  if (!(conf_level > 0.5 && conf_level < 1)) {
-    stop("conf_level should be greater than 0.5 and less than 1")
-  }
-
-  percentile_value <- conf_level + (1 - conf_level) / 2
-
-  conf_int <- function(y, conf_level, distribution) {
-    y <- stats::na.omit(y)
-
-    if (distribution == "normal") {
-      conf_int_out <- data.frame(
-        y = mean(y),
-        ymin = mean(y) - stats::qt(percentile_value,
-                                   length(y)) * sqrt(stats::var(y) / length(y)),
-        ymax = mean(y) + stats::qt(percentile_value,
-                                   length(y)) * sqrt(stats::var(y) / length(y))
+xgx_stat_ci <- function(
+                    data = NULL,
+                    mapping = NULL,
+                    geom = list("point", "line", "errorbar"),
+                    position = "identity",
+                    na.rm = FALSE,
+                    show.legend = NA, 
+                    inherit.aes = TRUE,
+                    
+                    conf_level = 0.95,
+                    distribution = "normal",
+                    bins = NULL,
+                    ...) {
+  lays <- list()
+  
+  for (igeom in geom) {
+    if (is.null(bins)) {
+      lay = layer(
+        stat = StatSummary,
+        data = data,
+        mapping = mapping,
+        geom = igeom,
+        position = position,
+        show.legend = show.legend,
+        inherit.aes = inherit.aes,
+        params = list(
+          fun.args = list(),
+          fun.data = function(y) xgx_conf_int(y, conf_level, distribution),
+          na.rm = na.rm,
+          ...)
       )
-    } else if (distribution == "lognormal") {
-      yy <- log(y)
-      conf_int_out <- data.frame(
-        y = exp(mean(yy)),
-        ymin = exp(mean(yy) - stats::qt(percentile_value, length(yy)) * sqrt(stats::var(yy) / length(yy))),
-        ymax = exp(mean(yy) + stats::qt(percentile_value, length(yy)) * sqrt(stats::var(yy) / length(yy)))
+    }
+    else {
+      lay = layer(
+        stat = StatCI,
```

Comment:
Maybe this should be a StatBin stat, that splits x into bins, and the fun.data can be customized like for StatSummary above

Reply:
I agree with this choice for simplicity, and will work on a solution that uses this.  

The reason I didn't do this from the start is the StatBin / StatSummaryBin objects do not calculate the bins that were desired by us (by quantile).  Although, different breaks can be used. So, a bins_by_quartile value could be provided, which could be used to calculate the breaks parameter within the xgx_stat_ci interface type of function, and then be given to the stat_summary_bin layer.

However, this would likely lose the ordinal functionality we have created within StatCI.

Reply:
I didn't mean to imply using the ggplot StatSummaryBin, for the exact reason that you stated - the bins were not calculated as we wanted. If you can figure out a way to either create our own StatBins or to use the StatSummaryBin breaking by quantiles, then either way should work.